### PR TITLE
Session rewrite: Generate pantheon.session and autostarts

### PR DESCRIPTION
**Pull request by worldofpeace**
**Map Pantheon required components by g-s-d versions**
To continue to support Pantheon with a newer g-s-d version in NixOS I needed to manually ship a different session file because of changes made in gnome-settings-daemon.
Let's make session-settings generate the pantheon.session with mappings to what components will be available in g-s-d by its version. This PR is inspired by what is already done in budgie-desktop. The minimum supported g-s-d version is >= 3.27.90, which is the GNOME 3.28 stack in the Ubuntu LTS. I will continue to make adjustments as new versions of g-s-d are released.

**Generate patched autostarts**
Keeping these in sync manually is completely unpractical.
What we do instead is use `desktop-file-install` to copy the autostarts from the respective package and edit the `OnlyShowIn` key.

### CODE REVIEW

**Comment by worldofpeace** [comment]
@SolarAquarion I believe `dev` will have the pkgconfig files which is used in the dependent commit of this PR.

**Comment by worldofpeace** [comment]
I'm seeing that the `gpg` autostart isn't even in gnome-keyring anymore. Dropping it.

**Comment by worldofpeace** [comment]
Here is the content of `user-dirs-update-gtk` autostart in eoan
```
[Desktop Entry]
Encoding=UTF-8
Exec=xdg-user-dirs-gtk-update
Name=User folders update
Comment=Update common folders names to match current locale
Terminal=false
NotShowIn=KDE;LXQt;
NoDisplay=true
Type=Application
StartupNotify=false
X-KDE-autostart-after=panel
X-Ubuntu-Gettext-Domain=xdg-user-dirs-gtk
```

no reason I can see that it wouldn't be ran.

Here is the patch
```
[Desktop Entry]
Encoding=UTF-8
Exec=xdg-user-dirs-gtk-update
Name=User folders update
Comment=Update common folders names to match current locale
Terminal=false
OnlyShowIn=Pantheon;
Type=Application
StartupNotify=false
X-KDE-autostart-after=panel
NoDisplay=true
X-Ubuntu-Gettext-Domain=xdg-user-dirs-gtk
```

I don't see why `OnlyShowIn` serves any purpose. I will remove this file.

File: release.nix [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{ nixpkgs ? <nixpkgs>
+}:
+
+let
+  pkgs = import nixpkgs { config = {}; overlays = []; };
+in rec {
```
Comment (jtojnar) [inline release.nix:6]: 1. Why clear the overlays?
2. Why not make the import the default argument?

```suggestion
{ nixpkgs ? import <nixpkgs> { }
}:

rec {
```

File: default.nix [hunk 2]
```diff
@@ -0,0 +1,47 @@
+{ stdenv
```
Comment (jtojnar) [inline default.nix:1]: It does not make much sense to me to have `default.nix` that is not instantiable.

File: session/meson.build [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+# Needs a minimum 3.28 GNOME stack
+gsd_minimum_version = '>= 3.27.90'
+
+gsd_dep = dependency('gnome-settings-daemon', version: gsd_minimum_version)
+gsd_version = gsd_dep.version()
+
+# Merge the list depending on the gnome-settings-daemon version.
+if gsd_version.version_compare('>=3.36.0')
+  gsd_components = gsd_components + gsd_usb_protection
+  session_components = pantheon_components + gsd_components
```
Comment (jtojnar) [inline session/meson.build:49]: This is in all three branches. Move it out?

File: release.nix (hunk 1 above)
Comment (worldofpeace) [reply release.nix]: I just generally think it's a good idea to clear overlays if I'm also going to build it with `nix-build`.  I've removed `release.nix` in latest commits.

File: default.nix [hunk 4]
```diff
@@ -0,0 +1,36 @@
+with import <nixpkgs> { config = {}; overlays = []; };
```
Comment (jtojnar) [inline default.nix:1]: This is now even less overridable. I would suggest making it parametrizable like here: https://gitlab.gnome.org/GNOME/buoh/-/blob/f37746cc3161f4b5fd4e9c88b725dfe2977cc400/default.nix#L24-33

File: session/meson.build [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+# Merge the list depending on the gnome-settings-daemon version.
+if gsd_version.version_compare('>=3.36.0')
+  gsd_components += gsd_usb_protection
+  session_components += gsd_components
+elif gsd_version.version_compare('>=3.33.90')
+  session_components += gsd_components
+else
+  gsd_components += gsd_mouse_clipboard
+  session_components += gsd_components
```
Comment (jtojnar) [inline session/meson.build:59]: This is still duplicate in every branch. I would suggest just using `session_components = pantheon_components + gsd_components` afterwards.

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -6,18 +6,30 @@ jobs:
...
+
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Nix
+        uses: cachix/install-nix-action@v8
+      - name: Build
+        run: 'nix-build --quiet default.nix -I nixpkgs=channel:20.03'
```
Comment (jtojnar) [inline .github/workflows/main.yml:35]:
```suggestion
        run: 'nix-build --quiet -I nixpkgs=channel:20.03'
```

**Comment by worldofpeace** [comment]
@danrabbit @cassidyjames @tintou Would like to see some signs of life on this pull request?

**Comment by tintou** [comment]
Thanks again

**Comment by decathorpe** [comment]
I've been doing something similar for fedora for years, I'll have to check whether I can drop my custom downstream thing in favor of this .... but glancing over the code I don't understand what it does yet :)

**Comment by worldofpeace** [comment]
@decathorpe Part of the motivation was to make it so you don't have that fork anymore :grimacing: 
I believe most everything should be here.

**Comment by decathorpe** [comment]
@worldofpeace the only problem I have right now is that Pantheon is currently broken on rawhide, so I can't begin to test this for fedora 33 :crying_cat_face:

**Pull request merged**